lib/CoreBot.js

```diff
@@ -40,13 +47,75 @@ function Botkit(configuration) {
     // define some middleware points where custom functions
     // can plug into key points of botkits process
     botkit.middleware = {
-        send: ware(),
-        receive: ware(),
         spawn: ware(),
-        heard: ware(),
+        ingest: ware(),
+        normalize: ware(),
+        categorize: ware(),
+        receive: ware(),
+        triggered: ware(), // triggered any event
+        heard: ware(), // best place for heavy i/o because fewer messages
         capture: ware(),
+        send: ware(),
+    };
+
+
+    botkit.ingest = function(bot, payload, source) {
+
+        // keep an unmodified copy of the message
+        payload.original_message = clone(payload);
+
+        payload._pipeline = {
+            stage: 'ingest',
+        };
+
+        // debug('ingest', payload);
+
+        botkit.middleware.ingest.run(bot, payload, source, function(err, bot, payload, source) {
+            botkit.normalize(bot, payload);
+        });
+    };
+
+    botkit.normalize = function(bot, payload) {
+        payload._pipeline.stage = 'normalize';
+        // debug('normalize', payload);
+        botkit.middleware.normalize.run(bot, payload, function(err, bot, message) {
+            if (!message.type) {
+                message.type = 'message_received';
+            }
+            debug('normalized', message);
```

Comment:
Long term I want to move to using the debug module. But, this was an accidental inclusion and I'll remove it for now.